semgrep-core/src/spacegrep/src/lib/Pre_match.ml

```diff
@@ -0,0 +1,73 @@
+(*
+   Fast check that tells whether a pattern has any chance of matching the
+   document with 'Match.search'.
+*)
+
+(*
+   Check if the pattern contains 'Byte' tokens, since they often don't
+   and the target often does. This is meant as an optimization to not
+   fill the lookup table with many useless entries.
+*)
+let pattern_has_bytes (pat : Pattern_AST.t) =
+  let open Pattern_AST in
+  let rec has_bytes = function
+    | Atom (_, atom) -> (
+        match atom with Byte _ -> true | Word _ | Punct _ | Metavar _ -> false )
+    | List nodes -> List.exists has_bytes nodes
+    | Dots _ | End -> false
+  in
+  List.exists has_bytes pat
+
+(* This is used when case-insensitive matching on words is requested. *)
+let lowercase = String.lowercase_ascii
+
+type atom_table = {
+  words : (string, unit) Hashtbl.t;
+  chars : (char, unit) Hashtbl.t; (* could use an array of 256 booleans *)
+}
+
+let build_atom_table ~case_sensitive (pat : Pattern_AST.t) =
+  let open Pattern_AST in
+  let words = Hashtbl.create 1000 in
+  let chars = Hashtbl.create 256 in
+  let rec index nodes = List.iter index_node nodes
+  and index_node = function
+    | Atom (_, atom) -> index_atom atom
+    | List nodes -> index nodes
+    | Dots _ | End -> ()
+  and index_atom = function
+    | Word word ->
+        let k = if case_sensitive then word else lowercase word in
+        Hashtbl.replace words k ()
+    | Punct k | Byte k -> Hashtbl.replace chars k ()
+    | Metavar _ -> ()
+  in
+  index pat;
+  { words; chars }
+
+(*
+   Warning: the atom table cannot be reused after this, since we operate
+   by removing the elements from the table.
+*)
+let doc_has_all_required_atoms ~case_sensitive ~check_bytes atom_tbl
+    (doc : Doc_AST.t) =
+  let open Doc_AST in
+  let { words; chars } = atom_tbl in
+  let rec check nodes = List.iter check_node nodes
+  and check_node = function
+    | Atom (_, atom) -> check_atom atom
+    | List nodes -> check nodes
+  and check_atom = function
+    | Byte k -> if check_bytes then Hashtbl.remove chars k
+    | Word word ->
+        let k = if case_sensitive then word else lowercase word in
+        Hashtbl.remove words k
+    | Punct k -> Hashtbl.remove chars k
+  in
+  check doc;
+  Hashtbl.length words = 0 && Hashtbl.length chars = 0
```

Comment:
This check is super cheap so we should be making it in `check` before going down into the nodes. Then you don't have to traverse the entire document in every case.

Reply:
maybe; it was just a first guess, I didn't check the impact but I don't think it's worth it since the whole module will be replaced by #3461.